fix: whiteout regex bug CN-272

parker-snyk opened the pull request:
We were using `.wh.` to look for whited out files, but we didn't escape the `.`, so it was a match-any character

We also weren't following the OpenContainer spec: https://github.com/opencontainers/image-spec/blob/main/layer.md#whiteouts

.wh. needs to be the prefix of the filename. So it either needs to prefix everything (no slashes, just filename) or it needs to prefix the filename after the last slash (full path)

Our current implementation just looks for .wh. to appear anywhere, so we are incorrectly whiting out files. This includes libraries, so all of the node 'which' packages were being skipped, for example

bgardiner commented on lib/extractor/index.ts:
```diff
@@ -228,7 +228,7 @@ function layersWithLatestFileModifications(
       // + not adding this file
       if (isWhitedOutFile(filename)) {
-        removedFilesToIgnore.add(filename.replace(/.wh./, ""));
+        removedFilesToIgnore.add(filename.replace(/\.wh\./, ""));
```
> I'll rework a regex to look for the start of the file name, but account for a file being in subdirectories

this makes sense to me

parker-snyk replied:
Yeah, it should be at the beginning of the filename. I'll fix this and the tests, and move the cleanup to a new pr

parker-snyk commented on lib/extractor/index.ts:
```diff
@@ -248,8 +248,32 @@ function layersWithLatestFileModifications(
+ * https://www.madebymikal.com/interpreting-whiteout-files-in-docker-image-layers
+ * https://github.com/opencontainers/image-spec/blob/main/layer.md#whiteouts
+ */
 export function isWhitedOutFile(filename: string) {
```
I think as long as it starts with .wh it should account for it, it looks like out logic checks if a file is or within any whited out file or directory, so if a directory starts with .wh, any files under it should be excluded. I can manually verify though

parker-snyk replied:
You were right, I didn't realize opaque means "nuke everything in this directory from lower layers." I've pushed a fix that adds isOpaqueWhiteout() detection and tracks opaque directories during layer merging so files from older layers in those dirs get properly excluded. Tests included

parker-snyk commented on lib/extractor/index.ts:
```diff
@@ -248,8 +248,32 @@ function layersWithLatestFileModifications(
+ */
 export function isWhitedOutFile(filename: string) {
-  return filename.match(/.wh./gm);
+  const lastSlashIndex = filename.lastIndexOf("/");
```
I updated to use path.basename so it should account for this now? https://nodejs.org/api/path.html#windows-vs-posix

parker-snyk committed:
fix: correct whiteout file detection and add opaque whiteout support (CN-272)

Two related bugs in OCI/Docker image layer whiteout handling:

* isWhitedOutFile previously matched any path containing the
  four-char pattern `?wh?` via an unescaped regex (/.wh./gm),
  silently dropping packages like which, whois, whitenoise, and
  python_whois from scan results. Anchor the check to a literal
  ".wh." basename prefix per the OCI Image Layer spec.

* No handling existed for opaque whiteout files (.wh..wh..opq).
  Per spec, an opaque whiteout in a newer layer means "delete
  everything in this directory from all older layers." Implement
  proper opaque-whiteout suppression scoped to older layers only.

New lib/extractor/whiteout.ts consolidates isWhitedOutFile,
isOpaqueWhiteout, and removeWhiteoutPrefix, with a getBasename
helper that handles both / and \\ separators for Windows scan
correctness. System test snapshots updated.
